src/distributions/categorical.rs

```diff
@@ -0,0 +1,279 @@
+use {Distribution, Generator};
+
+/// A discrete Categorical distribution.
+#[derive(Clone)]
+pub struct Categorical {
+    /// The size of the probabilty vector
+    pub k: usize,
+    /// The probability vector.
+    pub p: Vec<f64>,
+}
+
+fn is_prob_vec(p: &Vec<f64>) -> bool {
+    let mut sum = 0.;
+    for &p_i in p.iter() {
+        if p_i < 0.|| p_i > 1. { return false; }
+        sum += p_i;
+    }
+    (sum - 1.).abs() <= 1e-12
+}
+
+impl Categorical {
+    /// Create a Categorical distribution with success probability `p`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `p < 0` or `p > 1`.
+    #[inline]
+    pub fn new(p: Vec<f64>) -> Categorical {
+        debug_assert!(is_prob_vec(&p), "Categorical::new() is called with p not a probabilty vector");
+        Categorical { k: p.len() as usize,  p: p}
+    }
+
+    /// Create a Categorical distriburion with `k` categories of equal
+    /// probabilty `1/k`.
+    #[inline]
+    pub fn new_equal(k: usize) -> Categorical {
+        let p = 1. / k as f64;
+        Categorical { k: k,  p: vec![p; k] }
+    }
+}
+
+impl Distribution for Categorical {
+    type Value = usize;
```

Comment:
`usize` is here for easy array lookup. Since that is what I personally use the categorical distribution for most of the time.